query.py

```diff
@@ -0,0 +1,253 @@
+            fhir_api_url {str} -- The Service Base URL (e.g. http://hapi.fhir.org/baseR4/)
+
+        Keyword Arguments:
+            path_rules {str} -- path to the folder containing the CapabilityStatement and SearchParameters file (default: {None})
+        """
+        self.fhir_api_url = fhir_api_url
+        self.path_rules = path_rules
+
+        self.config = None
+        self.graph_query = None
+        self.dataframes = dict()
+        self.main_dataframe = None
+
+    def from_config(self, config: dict):
+        """Executes the query from a dictionary in the format of a configuration file
+
+        Arguments:
+            config {dict} -- dictionary in the format of a configuration file
+        """
+        self.config = {
+            "from_dict": config.get("from", None),
+            "select_dict": config.get("select", None),
+            "where_dict": config.get("where", None),
+            "join_dict": config.get("join", None),
+        }
+
+    def execute(self):
+        """Executes the complete query
+
+        1. constructs a GraphQuery object to store the query as a graph
+        2. builds the url of the requests to send to the API
+        3. retrieves the answers from the api and puts them as tables in the dataframes attribute
+        4. executes joins to return the result table in the main_dataframe attribute
+
+        Coming soon: cleaning unrequested columns
+        """
+        self.graph_query = GraphQuery(
+            fhir_api_url=self.fhir_api_url, path=self.path_rules
+        )
+        self.graph_query.execute(**self.config)
+        for (
+            resource_alias
+        ) in self.graph_query.resources_alias_info.keys():
+            elements = self.graph_query.resources_alias_info[
+                resource_alias
+            ]["elements"]
+
+            url_builder = URLBuilder(
+                fhir_api_url=self.fhir_api_url,
+                query_graph=self.graph_query,
+                main_resource_alias=resource_alias,
+            )
+
+            url = url_builder.search_query_url
+            call = ApiGetter(
+                url=url,
+                elements=elements,
+                main_resource_alias=resource_alias,
+            )
+            call.get_all()
+            self.dataframes[resource_alias] = call.display_data()
+        self._clean_columns()
+        self.main_dataframe = self._inner_join()
+        # to do check where
+        # clean to keep only wanted column
+
+    def _inner_join(self) -> pd.DataFrame:
+        """executes the joins one after the other in the order specified by the join_path function.
+
+        Returns:
+            pd.DataFrame -- dataframe containing all joined resources
+        """
+        list_join = join_path(
+            self.graph_query.resources_alias_graph
+        )
+        main_alias_join = list_join[0][0]
+        main_df = self.dataframes[main_alias_join]
+        for alias_1, alias_2 in list_join:
+            df_1 = main_df
```

Comment:
I'll try to answer it with an example :smile: 
Let's say that the join path is :  [(practitioner, allergy), (allergy, patient), (patient, condition)]
We initialize main_df with the `practitioner`-table.
The (practitioner, allergy) join has the effect of adding the `allergy` information in main_df.
Then at the join (allergy, patient), the information of `patient` is added to main_df.
Finally, the (patient, condition) join has the effect of adding the `condition` information in main_df.
We notice that at each step we were able to make the join because the key needed for the joint had been added to the main_df table before.
Thus, at the end of the join path, main_df contains all the information of the 4 tables concerned by the joins.

If we had put self.dataframes[alias_1 and not main_df, we would not have had a table at the end with all the information.

Did I answer your question correctly? Did I miss something? :smile: